### PR TITLE
HSolverDL subclass for backwards compat.

### PR DESCRIPTION
**Pull request by nwlambert**
Redoing this is a PR:
Added a HSolverDL subclass for backwards compatibility with the old version of the solver in qutip.  This should ideally mean anyone using the old solver should not be affected by the upgrades, though I did remove the stats and the renorm option.

### CODE REVIEW

**Comment by nwlambert** [comment]
Just for more info, I made a HSolverDL subclass of BosonicHEOMSolver. It accepts the same variables as the old HSolverDL in qutip,  but then uses them to construct the matsubara decomposition and sends them to the standard functions in BosonicHEOMSolver.  It seems to work ok (example 1a has a quick check), but I am not sure if I have done anything a bit silly in this sub-class approach, so perhaps @tehruhn or @quantshah  can just take a quick look.

**Comment by nwlambert** [comment]
Also added pure dephasing tests for both bosonicHEOMSolver and HSolverDL.  These are based on the old test alex wrote for the old HSolverDL.

**Comment by nwlambert** [comment]
Started on some slight modularization of both process_input() functions.

Also added a basic fermionic solver test based against the the single-impurity current solution (which was in example 4b).  This is kind of clunky because it needs lots of utility functions from that example.  I should just bite the bullet and include these in the heom.py

**Comment by nwlambert** [comment]
Thanks Shahnawaz.  if this inheritance stuff is ok, I might do another slight rearrangement,  with a parent class containing the .run() and .steadystate() functions, and have the boson and fermion things as sub-classes, and hsolverDL as a sub-sub-class.  Probably makes it easier in the long run.

File: tests/test_bosonic_solver.py [hunk 1]
```diff
@@ -3,23 +3,157 @@
...
+from bofin.heom import _heom_state_dictionaries, _check_Hsys
+import pytest
+from math import sqrt, factorial
 
+def test_add_at_idx():
+    """
+    Tests the function to add at hierarchy index.
+    """
+    seq = (2, 3, 4)
+    assert add_at_idx(seq, 2, 1), (2, 3, 5)
```
Comment (hodgestar) [inline tests/test_bosonic_solver.py:20]:
```suggestion
    assert add_at_idx(seq, 2, 1) == (2, 3, 5)
```
The syntax of assert is `assert <bool>, <error_msg>` so the current code asserts that `add_at_idx` returns a non-empty sequence and `(2, 3, 5)` is just the error message.

File: tests/test_bosonic_solver.py [hunk 2]
```diff
@@ -3,23 +3,157 @@
...
 
-from scipy.integrate import quad
+    seq = (2, 3, 4)
+    assert add_at_idx(seq, 0, -1), (1, 3, 4)
```
Comment (hodgestar) [inline tests/test_bosonic_solver.py:23]:
```suggestion
    assert add_at_idx(seq, 0, -1) == (1, 3, 4)
```

File: tests/test_bosonic_solver.py [hunk 3]
```diff
@@ -3,23 +3,157 @@
...
+        )
+    assert nhe, total_nhe
+        
+def test_check_H():
+    """Tests the function for checking system Hamiltonian"""
+    assert(_check_Hsys(sigmax()))
+    assert(_check_Hsys([sigmax(), sigmaz()]))
+    assert(_check_Hsys([[sigmax(), np.sin], [sigmaz(), np.cos]]))
+    assert(_check_Hsys([[sigmax(), np.sin], [sigmaz(), np.cos]]))
+    assert(_check_Hsys(QobjEvo([sigmaz(), sigmax(), sigmaz()])))
```
Comment (hodgestar) [inline tests/test_bosonic_solver.py:50]:
```suggestion
    _check_Hsys(sigmax())
    _check_Hsys([sigmax(), sigmaz()])
    _check_Hsys([[sigmax(), np.sin], [sigmaz(), np.cos]])
    _check_Hsys([[sigmax(), np.sin], [sigmaz(), np.cos]])
    _check_Hsys(QobjEvo([sigmaz(), sigmax(), sigmaz()]))
```
The `_check_Hsys` function should either return `None` and raise an exception if the check fails, or return `True` or `False`. Currently it returns `True` and raises an exception which mixes the two patterns in a way that makes the returned value redundant -- if `_check_Hsys` returns at all the check has already passed and the return value is irrelevant.

File: tests/test_bosonic_solver.py [hunk 4]
```diff
@@ -3,23 +3,157 @@
...
+    H_sys = Qobj(np.zeros((2, 2)))
+    Q = sigmaz()
+    initial_state = 0.5*Qobj(np.ones((2, 2)))
+    projector = basis(2, 0) * basis(2, 1).dag()
+    options = Options(nsteps=15_000, store_states=True)
+    hsolver = HSolverDL(H_sys, Q, coupling_strength, temperature,
+                        14, 2, cut_frequency,
+                        bnd_cut_approx=bnd_cut_approx,
+                        options=options)
+    test = expect(hsolver.run(initial_state, times).states, projector)
```
Comment (hodgestar) [inline tests/test_bosonic_solver.py:107]: Since we're testing the solver results we should probably assert things about the rest of the result too:
```
results = hsolver.run(...)
expected = expect(result.states, projector)
np.testing.assert_allclose(test, expected, atol=tol)
... assert things about other parts of the result ...
```

File: tests/test_bosonic_solver.py [hunk 5]
```diff
@@ -3,23 +3,157 @@
...
+    np.testing.assert_allclose(test, expected, atol=tol)
+
+
+@pytest.mark.filterwarnings("ignore::scipy.integrate.IntegrationWarning")
+def test_pure_dephasing_model_BosonicHEOMSolver():
+    """
+    BosonicHEOMSolver: Compare with pure-dephasing analytical assert that the
+    analytical result and HEOM produce the same time dephasing evoltion.
+    """
+    tol=1e-3
```
Comment (hodgestar) [inline tests/test_bosonic_solver.py:118]:
```suggestion
    tol = 1e-3
```

File: tests/test_fermionic_solver.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
+"""
+Tests for the Bosonic HEOM solvers.
```
Comment (hodgestar) [inline tests/test_fermionic_solver.py:2]:
```suggestion
Tests for the Fermionic HEOM solvers.
```

File: tests/test_fermionic_solver.py [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+        )
+    assert nhe, total_nhe
+        
+  
+@pytest.mark.filterwarnings("ignore::scipy.integrate.IntegrationWarning")
+def test_discrete_level_model_FermionicHEOMSolver():
+    """
+    FermionicHEOMSolver: compare to discrete-level current analytics
+    """
+    tol=1e-3
```
Comment (hodgestar) [inline tests/test_fermionic_solver.py:42]:
```suggestion
    tol = 1e-3
```

File: tests/test_fermionic_solver.py [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+    Gamma = 0.01  #coupling strength
+    W=1. #cut-off
```
Comment (hodgestar) [inline tests/test_fermionic_solver.py:44]:
```suggestion
    W = 1. #cut-off
```

File: tests/test_fermionic_solver.py [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+    Qops = [d1.dag(),d1,d1.dag(),d1]
+
+
+
+    Kk=lmax+1
+    Ncc = 2  #For a single impurity we converge with Ncc = 2
+    
+    eta_list = [etapR,etamR,etapL,etamL]
+    gamma_list = [gampR,gammR,gampL,gammL]
+    Qops = [d1.dag(),d1,d1.dag(),d1]
```
Comment (hodgestar) [inline tests/test_fermionic_solver.py:176]: This seems to overwrite the previous definition of `Qops` with the same value it had?

File: tests/test_fermionic_solver.py [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+        for i in aux_indices:
+            qlist = [Qobj(full[k, i, :].reshape(shape, shape).T,dims=dims) for k in range(len(full))]
+            aux.append(qlist)
+        return aux, aux_heom_indices, idx2state
+    
+    aux_1_list_list=[]
+    aux1_indices_list=[]
+    aux_2_list_list=[]
+    aux2_indices_list=[]
+
```
Comment (hodgestar) [inline tests/test_fermionic_solver.py:231]:
```suggestion
```
These are set a few lines down, so no need to set them here.

**Review by hodgestar — COMMENTED** [review]
Some initial review comments.

I would like to rewrite `_check_Hsys` again :) -- the current recursive structure is incorrect. The lists allowed for `H` cannot be arbitrarily nested, so the allowed cases will be much clearer if the function doesn't call itself.

File: tests/test_bosonic_solver.py (hunk 3 above)
Comment (nwlambert) [reply tests/test_bosonic_solver.py:50]: yeah, this sounds sensible!

File: tests/test_bosonic_solver.py (hunk 4 above)
Comment (nwlambert) [reply tests/test_bosonic_solver.py:107]: I guess we can just check the output is normalized at least, there's not much else we can do with this particular example I think.

File: tests/test_fermionic_solver.py (hunk 9 above)
Comment (nwlambert) [reply tests/test_fermionic_solver.py:176]: just making sure it sticks :)

**Comment by hodgestar** [comment]
@nwlambert @quantshah I've started merging BoFiN and this PR into QuTiP at qutip/qutip#1601. Comments and reviews there appreciated.
